docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,322 @@
+
+Pull a repository and display its metadata as formatted JSON in standard output. The following fields should be formatted in a JSON output:
+
+- `Ref`: full artifact reference by digest, e.g, `$REGISTRY/$REPO@$DIGEST`
+-  `Files`: a list of downloaded files
+    -  `Path`: the absolute file path of the pulled file (layer)
+    -  `Ref`: full reference by digest of the pulled file (layer)
+    -  `MediaType`: media type of the pulled file (layer) 
+    -  `Digest`: digest of the pulled file (layer) 
+    -  `Size`: file size in bytes
+
+Pull a repository that contains multiple layers (files) and show their descriptor metadata as pretty JSON in standard output.
+
+```bash
+oras pull $REGISTRY/$REPO:$TAG --artifact-type example/sbom sbom.spdx  --artifact-type example/vul-scan vul-scan.json --format json
+```
+
+```json
+{
+  "Ref": "localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186111",
+  "Files": [
+    {
+      "Path": "/home/user/oras-install/sbom.spdx",
+      "Ref": "localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186222",
+      "MediaType": "application/vnd.oci.image.manifest.v1+json",
+      "Digest": "sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186222",
+      "Size": 820,
+      "Annotations": {
+        "org.opencontainers.image.title": "sbom.spdx"
+      }
+    },
+    {
+      "Path": "/home/user/oras-install/vul-scan.json",
+      "Ref": "localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd18669b",
+      "MediaType": "application/vnd.oci.image.manifest.v1+json",
+      "Digest": "sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd18669b",
+      "Size": 820,
+      "Annotations": {
+        "org.opencontainers.image.title": "vul-scan.json"
+      }
+    }
+  ]
+}
+```
+
+Pull a repository that contains multiple layers (files) and display its descriptor metadata as raw JSON in standard output.
+
+```bash
+oras pull $REGISTRY/$REPO:$TAG --format '{{toRawJson .}}'
+```
+
+```
+{"Ref":"localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186111","Files":[{"Path":"/home/user/oras-install/sbom.spdx","Ref":"localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186222","MediaType":"application/vnd.oci.image.manifest.v1+json","Digest":"sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186222","Size":820},{"Path":"/home/user/oras-install/vul-scan.json","Ref":"localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd18669b","MediaType":"application/vnd.oci.image.manifest.v1+json","Digest":"sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd18669b","Size":820}]}
+```
+
+### oras push
+
+Push two files to a repository and show the descriptor of the image manifest in pretty JSON format.
+
+```bash
+oras push $REGISTRY/$REPO:$TAG sbom.spdx vul-scan.json --format json 
+```
+
+```json
```

Comment:
What if OCI image spec add a field named `ref` to descriptor in the future?

**I suggest we pick out all possible properties of a descriptor to a dedicated field like `Descriptor`**(name TBD).
```json
{
  "Ref": "localhost:5000/oras@sha256:4a5b8c83d153f52afdfcb422db56c2349aae3bd5ecf8338a58353b5eb6681c45",
  "Descriptor": {
    "MediaType": "application/vnd.oci.image.manifest.v1+json",
    "Digest": "sha256:4a5b8c83d153f52afdfcb422db56c2349aae3bd5ecf8338a58353b5eb6681c45",
    "Size": 820,
    "Annotations": {
      "org.opencontainers.image.created": "2023-12-15T09:41:54Z"
    }
  }
}
```
This is to make this design upward-compatible to any OCI image spec changes. If the above issue does happen, we can simply add `Ref` to `Descriptor`.

Reply:
This can be also applied to `oras pull` and `oras attach`

Reply:
Good suggestion. Should we use `Config` as the field name here?

Reply:
No, it's the descriptor of generated manifest. If we need to add another descriptor for `--config`, we should name it as Config.

Reply:
Resolving go-template selectable fields are not expected to be compatible with image spec.